context-propagation/src/test/java/io/micrometer/context/DefaultContextSnapshotTests.java

```diff
@@ -53,6 +53,24 @@ void should_propagate_thread_local() {
         }
     }
 
+    @Test
+    void test_null_value_in_source_context() {
+        this.registry.registerContextAccessor(new TestContextAccessor());
+        this.registry.registerThreadLocalAccessor(new ObservationThreadLocalAccessor());
+
+        String key = ObservationThreadLocalAccessor.KEY;
+
+        String emptyValue = ObservationThreadLocalHolder.getValue();
+        Map<String, String> sourceContext = Collections.singletonMap(key, emptyValue);
+
+        ContextSnapshot snapshot = ContextSnapshot.captureAll(this.registry, sourceContext);
+
+        try (Scope scope = snapshot.setThreadLocals()) {
+            assertThat(ObservationThreadLocalHolder.getValue()).isEqualTo(emptyValue);
+        }
+        assertThat(ObservationThreadLocalHolder.getValue()).isEqualTo(emptyValue);
+    }
+
```

Comment:
This test method could go further down in the overall order. Right now it's in the middle of `should_propagate_*` tests. Perhaps, it can go right after `should_filter_*` tests.

It would also be useful to have several tests around this, to cover all the ways in which a `null` value could potentially make its way in:

1. capturing from `ThreadLocal` values into a `ContextSnapshot`
2. capturing from a context into a `ContextSnapshot (covered)
3. setting all `ThreadLocal` values directly from a context
4. setting a subset of `ThreadLocal` values from a context.

For the implementation, one thought is to downcast the `ContextSnapshot` to a `Map` since we know that's what is created by default, and check for keys. Seems more transparent and easier to understand, and such an assumption in tests would be okay.

Reply:
`should_not_fail_on_empty_thread_local` addresses point 1. I will move it closer to the other tests around null situations.
For 3 and 4 I intend to validate in a single test by calling `ContextSnapshot.setAllThreadLocals` and afterwards `ContextSnapshot.setThreadLocals` with a key.